core/src/main/java/org/apache/calcite/sql/type/BasicSqlType.java

```diff
@@ -180,8 +180,10 @@ protected void generateTypeString(StringBuilder sb, boolean withDetail) {
     boolean printScale = scale != SCALE_NOT_SPECIFIED;
 
     // for the digest, print the precision when defaulted,
-    // since (for instance) TIME is equivalent to TIME(0).
-    if (withDetail) {
+    // but for TIME or TIMESTAMP, although (for instance)
+    // TIME is equivalent to TIME(0), the digest of TIME
+    // and TIME(0) should be different literally.
+    if (withDetail && !SqlTypeName.DATETIME_TYPES.contains(typeName)) {
       // -1 means there is no default value for precision
```

Comment:
The whole if block should be removed. Not only the `DATETIME_TYPES ` has this problem, all the types that allows precision could have such a problem, the root cause is that `#toString` and `digest` has different handling of printing non specified precision.

One way to solve is to not fire caching of types without precision and with default precision.

Reply:
The code is very old, back to the first commit in 2012, considering minimum change, I add this check. Disabling `canonize` for some types with conditions would also add more logic. So I am thinking that removing the whole condition block (all unit tests will pass) would be a good solution.

Reply:
comment addressed.